types/time.go

```diff
@@ -2980,76 +2980,157 @@ func minutesNumeric(t *CoreTime, input string, ctx map[string]int) (string, bool
 	return input[length:], true
 }
 
-const time12HourLen = len("hh:mm:ssAM")
+type parseState int32
 
-func time12Hour(t *CoreTime, input string, ctx map[string]int) (string, bool) {
-	// hh:mm:ss AM
-	if len(input) < time12HourLen {
-		return input, false
-	}
-	hour, succ := parseDigits(input, 2)
-	if !succ || hour > 12 || hour == 0 || input[2] != ':' {
-		return input, false
+const (
+	parseStateNormal    parseState = 1
+	parseStateFail      parseState = 2
+	parseStateEndOfLine parseState = 3
+)
+
+func parseSep(input string) (string, parseState) {
+	input = skipWhiteSpace(input)
+	if len(input) == 0 {
+		return input, parseStateEndOfLine
 	}
-	// 12:34:56 AM -> 00:34:56
-	if hour == 12 {
-		hour = 0
+	if input[0] == ':' {
+		return input[1:], parseStateNormal
 	}
+	return input, parseStateFail
+}
 
-	minute, succ := parseDigits(input[3:], 2)
-	if !succ || minute > 59 || input[5] != ':' {
-		return input, false
-	}
+func time12Hour(t *CoreTime, input string, ctx map[string]int) (string, bool) {
+	tryParse := func(input string) (string, parseState) {
+		state := parseStateNormal
+		// hh:mm:ss AM
+		/// Note that we should update `t` as soon as possible, or we
+		/// can not get correct result for incomplete input like "12:13"
+		/// that is shorter than "hh:mm:ss"
+		result := oneOrTwoDigitRegex.FindString(input) // 1..12
+		length := len(result)
+		hour, succ := parseDigits(input, length)
+		if !succ || hour > 12 || hour == 0 {
+			return input, parseStateFail
+		}
+		// Handle special case: 12:34:56 AM -> 00:34:56
+		// For PM, we will add 12 it later
+		if hour == 12 {
+			hour = 0
+		}
+		t.setHour(uint8(hour))
 
-	second, succ := parseDigits(input[6:], 2)
-	if !succ || second > 59 {
-		return input, false
+		if input, state = parseSep(input[length:]); state != parseStateNormal {
+			return input, state
+		}
+
+		if input = skipWhiteSpace(input); len(input) == 0 {
+			return input, parseStateEndOfLine
+		}
```

Comment:
I think add skipWhiteSpace into parseSep when state is parseStateNormal is better.  So the input can used by the next digit parse.

Reply:
And why need three states?   I think fail or succ is enough..

Reply:
> why need three states? I think fail or succ is enough..

Because MySQL accepts parsing some inputs like "11:13" with format "%r", which can be parsed into a datetime "11:13:00".

In `parSep`, we may run into three situations:
1. meet an ":", then we should accept it and continue to parse string left
2. meet a char that is not ":", then we should stop the parsing and return a "ZeroTime"
3. meet the end of the input string, then we should stop the parsing, with some parsing result

There are three different following actions, so we need three different states.

Reply:
> add skipWhiteSpace into parseSep when state is parseStateNormal

Address